### PR TITLE
QA-SRL

### PR DESCRIPTION
**Pull request by zphang**
Re: #571

Starting with a basic implementation of QA-SRL, seeking feedback.

- There is quite a lot of repeated/branching input sets in QA-SRL. Each sentence has multiple "verbs", each verb corresponds to multiple questions, each question has multiple answers from different annotators, and each answer contains potentially more than one span. Currently I'm doing a naive implementation of enumerating *all* of these. For a sense of scale, there are about 44k sentences in the training set, but about 600k+ total training examples when expanded.
- The QA-SRL / Large-Scale QA-SRL papers did not talk too much about scoring, and where they did I'm not sure if it works best for pretraining. I'm currently implementing the following:
- **Training loss**: I am currently following the BERT formulation for SQuAD of straightforwardly predicting the span start and ends via a softmax over the number of tokens in the input, with a cross-entropy loss. No correction for ensuring start < end.
- **Evaluation**: I am currently using the SQuAD-style macro-F1, which treats the predicted / gold tokens as individual observations, computing the F1 for each example, and then averaging across examples.

### CODE REVIEW

**Comment by sleepinyourhat** [comment]
This all sounds reasonable. You might get some odd behavior with SQuAD-style prediction, since it's not really well defined with uncertainty: When the model is predicting the end of a span, it doesn't know which span it predicted the start for. But I think this is a reasonable setup to start with.

**Comment by zphang** [comment]
Updated with pred writing logic.

One (maybe minor) issue is that the data instances constructed from expanding the questions retain

**Comment by sleepinyourhat** [comment]
Hrm. Do you see an easy way to shuffle the examples when loading them? 

I know the loading code is a bit hairy, so if not, mind just adding a warning in the loader that says not to use val_data_limit?

**Comment by zphang** [comment]
Using val_data_limit is probably still recommended, since the validation set is quite large otherwise.

I could simply add a fixed-seed random shuffle while loading the validation set. That way the order will always be consistent.

This issue may affect others tasks that create instances by expanding inputs (e.g. MultiRCTask).

**Comment by sleepinyourhat** [comment]
Fixed shuffle sounds good to me!

**Comment by sleepinyourhat** [comment]
(Reviewing remaining changes now.)

**Comment by zphang** [comment]
+ Or actually any task that has an order to the validation data, if the examples aren't shuffled somewhere along the way.

https://github.com/nyu-mll/jiant/blob/3cab32df250043c290846cc03f1b6187da897c19/src/trainer.py#L832-L838

File: src/modules/modules.py [hunk 1]
```diff
@@ -0,0 +1,1011 @@
+""" Different model components to use in building the overall model.
```
Comment (sleepinyourhat) [inline src/modules/modules.py]: This looks like a merge conflict—we deleted most of this file in a restructuring earlier in the week.

File: src/tasks/qa.py [hunk 2]
```diff
@@ -172,3 +177,143 @@ def get_metrics(self, reset=False):
...
 
         return {"ans_f1": ans_f1, "qst_f1": qst_f1, "em": em, "avg": (ans_f1 + em) / 2}
+
+
+@register_task("qasrl", rel_path="QASRL/")
+class QASRLTask(SpanPredictionTask):
+    def __init__(self, path, max_seq_len, name, **kw):
+        """QA-SRL (Question-Answer Driven Semantic Role Labeling)
+        See http://qasrl.org/
+        Move the contents of qasrl-v2 into QASRL
```
Comment (sleepinyourhat) [inline src/tasks/qa.py]: ?

File: src/tasks/qa.py [hunk 3]
```diff
@@ -172,3 +177,143 @@ def get_metrics(self, reset=False):
...
+        split = list(split)
+        instances = map(_make_instance, *split)
+        return instances
+
+    def _load_file(self, path):
+        example_list = []
+        aligner_fn = get_aligner_fn(self.tokenizer_name)
+        with gzip.open(path) as f:
+            lines = f.read().splitlines()
+            for line in tqdm.tqdm(lines):
```
Comment (sleepinyourhat) [inline src/tasks/qa.py]: Probably best to drop for consistency—we don't do this anywhere else.

File: src/allennlp_mods/span_metrics.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        f1 = 2 * precision * recall / (precision + recall)
+
+        # F1 is 0 is there is no overlap
+        f1[np.isnan(f1)] = 0
+
+        self._f1_total += f1.sum(0)
+        self._count += len(pred_start)
+
+    def get_metric(self, reset: bool = False):
+        f1 = self._f1_total / self._count
```
[the remark this answers is not in the document]
Comment (zphang) [reply src/allennlp_mods/span_metrics.py]: I believe SQuaD evaluation uses per-example F1. I also referenced this: https://github.com/allenai/bi-att-flow/blob/49004549e9a88b78c359b31481afa7792dbb3f4a/squad/evaluate.py#L29-L39

File: src/evaluate.py [hunk 5]
```diff
@@ -153,6 +165,7 @@ def write_preds(
             + ["wmt"]
             + tasks_module.ALL_COLA_NPI_TASKS
         )
+        simple_csv_preds_tasks = ["qasrl"]
```
[the remark this answers is not in the document]
Comment (zphang) [reply src/evaluate.py]: Adding it as a static attribute of tasks may not be a good idea: we would either have to define that attribute for all tasks (including tasks which already have specific prediction-writing logic defined), or have to use a `hasattr` which is quite ugly. 

Since `evaluate.py` has checks against task classes, I could alternatively add
```
SIMPLE_PRED_CSV_TASKS = [QASRLTask]
```
as a constant to the file.

File: src/tasks/qa.py (hunk 2 above)
Comment (zphang) [reply src/tasks/qa.py]: Rephrasing to:

```Download, unzip, and rename the "qasrl-v2" folder to "QASRL"```

File: src/evaluate.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (sleepinyourhat) [reply src/evaluate.py]: IIRC, all tasks inherit from the same class. How about we add csv_preds = False to that class, and override it here? I think Yada does something similar in another PR for another task attribute.

File: src/tasks/qa.py (hunk 2 above)
Comment (sleepinyourhat) [reply src/tasks/qa.py]: SGTM. Wasn't sure if this was an instruction for the reader or a TODO.

**Comment by zphang** [comment]
Updated after feedback.

**Comment by sleepinyourhat** [comment]
@zphang Are you still planning to merge this? It'd definitely be helpful, but there's more to do to get it up to date. Otherwise close.

**Comment by zphang** [comment]
I'll take a look to get it up to date.

**Comment by sleepinyourhat** [comment]
@pruksmhc - Do you have time to take a look at the modeling side?

File: jiant/tasks/tasks.py [hunk 6]
```diff
@@ -198,6 +198,8 @@ class Task(object):
         - optimizer
     """
 
+    CSV_PREDS = False
```
Comment (sleepinyourhat) [inline jiant/tasks/tasks.py]: Add a comment explaining this.

File: jiant/tasks/tasks.py [hunk 7]
```diff
@@ -2570,6 +2572,13 @@ class MultipleChoiceTask(Task):
     pass
 
 
+class SpanPredictionTask(Task):
+
```
Comment (sleepinyourhat) [inline jiant/tasks/tasks.py]: Odd spacing, and docstring is usually the first line.

File: jiant/utils/utils.py [hunk 8]
```diff
@@ -180,12 +180,19 @@ def parse_json_diff(diff):
     actual value of the replaced or inserted item, whereas for jsondiff.delete, we do not want to
     show deletions in our parameters.
     For example, for jsondiff.replace, the output of jsondiff may be the below:
+<<<<<<< HEAD:jiant/utils/utils.py
```
Comment (sleepinyourhat) [inline jiant/utils/utils.py]: Fix.

File: jiant/tasks/qa.py [hunk 9]
```diff
@@ -423,3 +428,160 @@ def get_metrics(self, reset=False):
...
         return {"f1": f1, "em": em, "avg": (f1 + em) / 2}
+
+
+@register_task("qasrl", rel_path="QASRL/")
+class QASRLTask(SpanPredictionTask):
+
+    CSV_PREDS = True
+
+    def __init__(self, path, max_seq_len, name, **kw):
+        """QA-SRL (Question-Answer Driven Semantic Role Labeling)
```
Comment (sleepinyourhat) [inline jiant/tasks/qa.py]: Make a PR to the documentation (jiant-site repo) that mentions that this task is supported, and how to download the data.

File: src/tasks/qa.py [hunk 10]
```diff
@@ -0,0 +1,326 @@
+"""Task definitions for question answering tasks."""
```
Comment (sleepinyourhat) [inline src/tasks/qa.py]: Do you know why this is showing up as a diff? Most of this isn't new, right?

Comment (sleepinyourhat) [reply src/tasks/qa.py]: Ah, I see. Your changes are in src/, but the files in the repo are in jiant/. You'll have to move your changes over manually.

File: jiant/tasks/tasks.py (hunk 6 above)
Comment (zphang) [reply jiant/tasks/tasks.py]: Removed, and replaced with a direct check for task class.

File: jiant/utils/utils.py (hunk 8 above)
Comment (zphang) [reply jiant/utils/utils.py]: Fixed.

File: jiant/tasks/qa.py (hunk 9 above)
Comment (zphang) [reply jiant/tasks/qa.py]: Submitted.

File: src/tasks/qa.py (hunk 10 above)
Comment (zphang) [reply src/tasks/qa.py]: Should be fixed now.

**Comment by sleepinyourhat** [comment]
Could you update or close this this this week?

**Comment by zphang** [comment]
Will do.

File: jiant/allennlp_mods/span_metrics.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
+import numpy as np
+import torch
+
+from allennlp.training.metrics.metric import Metric
+
+
+@Metric.register("spanf1")
+class SpanF1Measure(Metric):
```
Comment (W4ngatang) [inline jiant/allennlp_mods/span_metrics.py]: any reason not to use the allennlp version? 

https://github.com/allenai/allennlp/blob/master/allennlp/training/metrics/span_based_f1_measure.py

File: jiant/tasks/qa.py [hunk 12]
```diff
@@ -441,3 +446,165 @@ def get_metrics(self, reset=False):
...
+        self.sentences = (
+            self.train_data[0] + self.train_data[1] + self.val_data[0] + self.val_data[1]
+        )
+
+    def get_sentences(self) -> Iterable[Sequence[str]]:
+        """ Yield sentences, used to compute vocabulary. """
+        yield from self.sentences
+
+    def process_split(
+        self, split, indexers, model_preprocessing_interface
```
Comment (W4ngatang) [inline jiant/tasks/qa.py]: what is `model_preprocessing_interface`?

File: jiant/tasks/qa.py [hunk 13]
```diff
@@ -441,3 +446,165 @@ def get_metrics(self, reset=False):
...
+        return [
+            [example[k] for example in example_list]
+            for k in ["sentence_tokens", "question_tokens", "answer_span", "idx"]
+        ]
+
+    @classmethod
+    def _shuffle_data(cls, data, seed=1234):
+        # Shuffle validation data to ensure diversity in periodic validation with val_data_limit
+        indices = list(range(len(data[0])))
+        random.Random(seed).shuffle(indices)
```
Comment (W4ngatang) [inline jiant/tasks/qa.py]: what's with the seed here? Is it not enough to seed `random` in main ?

File: jiant/evaluate.py [hunk 14]
```diff
@@ -473,6 +487,11 @@ def _write_rte_preds(
             preds_fh.write("{0}\n".format(json.dumps(out_d)))
 
 
+def _write_simple_csv_preds(task, preds_df: pd.DataFrame, pred_dir: str, split_name: str):
```
Comment (W4ngatang) [inline jiant/evaluate.py]: tsv ?

File: jiant/modules/simple_modules.py [hunk 15]
```diff
@@ -234,3 +234,30 @@ def forward(self, s1, s2, mask1, mask2, idx1=[], idx2=[]):
         pair_emb = torch.cat([emb1, emb2, torch.abs(emb1 - emb2), emb1 * emb2], 1)
         logits = self.classifier(pair_emb)
         return logits
+
+
+class TokenProjectionEncoder(nn.Module):
```
Comment (W4ngatang) [inline jiant/modules/simple_modules.py]: Just checking my understanding: `TokenProjectionEncoder` applies the same linear projections to each vec in a sequence of vectors, and `TokenMultiProjectionEncoder` applies multiple linear projections to each vec (like multi-head attention) ?

Are you normally setting `d_proj << d_inp` ? They're the same by default

File: jiant/tasks/qa.py (hunk 13 above)
Comment (W4ngatang) [reply jiant/tasks/qa.py]: Is this being called somewhere right now ?

**Comment by zphang** [comment]
In that case, should we merge and I'll add/assign an issue to myself to revisit the evaluation?

**Comment by zphang** [comment]
Wait, sorry, just saw the comments on the commit. Let me address them.

File: jiant/tasks/qa.py (hunk 12 above)
Comment (zphang) [reply jiant/tasks/qa.py]: It seems to be a new addition following the new integration of pytorch_transformers, for handling the different model types.

File: jiant/tasks/qa.py (hunk 13 above)
Comment (zphang) [reply jiant/tasks/qa.py]: See: https://github.com/nyu-mll/jiant/pull/716#issuecomment-506966644

The goal is to have a fixed-order shuffle of the validation data, potentially across runs. Hence, we want it to be independent of the usual random seed.

File: jiant/evaluate.py (hunk 14 above)
Comment (zphang) [reply jiant/evaluate.py]: Fixed!

File: jiant/modules/simple_modules.py (hunk 15 above)
Comment (zphang) [reply jiant/modules/simple_modules.py]: After taking a look, I've refactored this to make it based on a dict of classifiers.

File: jiant/allennlp_mods/span_metrics.py (hunk 11 above)
Comment (zphang) [reply jiant/allennlp_mods/span_metrics.py]: It looks like AllenNLP has two options. The one you linked which is based on Conll SRL, which is based on BIO (+variants) style of labels, and https://github.com/allenai/allennlp/blob/master/allennlp/training/metrics/squad_em_and_f1.py for SQuAD, but is based on raw strings. Neither is quite what we are looking for (token-based EM/F1 for spans). Arguably we can reformulate the labels to match the BIO variant (I'm not too familiar with this), but let me know what you think.

File: jiant/modules/simple_modules.py (hunk 15 above)
Comment (zphang) [reply jiant/modules/simple_modules.py]: Actually, decided to go with removed the unnecessary projection layer. Pushed update.

**Comment by zphang** [comment]
I think this is good to merge.

**Comment by sleepinyourhat** [comment]
Cool, do add the issue, though, if this wasn't fixed.
